src/finalizer/finalizer.cpp

```diff
@@ -5,207 +5,325 @@
+                        hr = WiuConfigureProductEx(sczProductId, INSTALLLEVEL_DEFAULT, INSTALLSTATE_ABSENT, L"MSIFASTINSTALL=7 IGNOREDEPENDENCIES=ALL REBOOT=ReallySuppress", &restart);
+                        LogStringLine(REPORT_STANDARD, "  Uninstall of \"%ls\" (%ls%) exited with 0x%.8x", sczProductName, sczProductId, hr);
+
+                        // Flag any reboot since we need to return that to the bundle.
+                        if (WIU_RESTART_INITIATED == restart || WIU_RESTART_REQUIRED == restart)
+                        {
+                            LogStringLine(REPORT_STANDARD, "  Reboot requested, deferring.");
+                            *pbRestartRequired = TRUE;
+                        }
+
+                        // Reset potential failures so we can continue to remove as many dependents as possible.
+                        hr = S_OK;
+                    }
+                    else if (HRESULT_FROM_WIN32(ERROR_UNKNOWN_PRODUCT) == hr || HRESULT_FROM_WIN32(ERROR_UNKNOWN_PROPERTY) == hr)
+                    {
                         // Possibly a corrupted provider key that wasn't cleaned up. We'll just ignore it.
-						LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
-						hr = S_OK;
-					}
-				}
-			}
-		}
+                        LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
+                        hr = S_OK;
+                    }
+                }
+            }
+        }
 
-		ReleaseRegKey(hkDependentsKey);
-		ReleaseRegKey(hkProviderKey);
-	}
+        ReleaseRegKey(hkDependentsKey);
+        ReleaseRegKey(hkProviderKey);
+    }
 
 LExit:
-	ReleaseStr(sczProductName);
-	ReleaseStr(sczProductId);
-	ReleaseStr(sczProviderKey);
-	ReleaseStr(sczDependentsKey);
-	ReleaseRegKey(hkDependentsKey);
-	ReleaseRegKey(hkProviderKey);
-	ReleaseRegKey(hkInstallerDependenciesKey);
-	return hr;
+    ReleaseStr(sczProductName);
+    ReleaseStr(sczProductId);
+    ReleaseStr(sczProviderKey);
+    ReleaseStr(sczDependentsKey);
+    ReleaseRegKey(hkDependentsKey);
+    ReleaseRegKey(hkProviderKey);
+    ReleaseRegKey(hkInstallerDependenciesKey);
+    return hr;
+}
+
+extern "C" HRESULT ParseSdkVersion(LPWSTR sczSdkVersion, INT * piMajor, INT * piMinor, INT * piFeatureBand)
+{
+    HRESULT hr = S_OK;
+    UINT cVersionParts = 0;
+    LPWSTR* rgsczVersionParts = NULL;
+
+    hr = StrSplitAllocArray(&rgsczVersionParts, &cVersionParts, sczSdkVersion, L".");
+    ExitOnFailure(hr, "Failed to split version.");
+
+    // We only care about the major.minor.patch values
+    // to convert to a feature band. If we don't have at least
+    // all 3 parts, we'll ignore the value.
+    if (3 > cVersionParts)
+    {
+        ExitOnFailure(E_INVALIDARG, "Invalid SDK version: %ls %li", sczSdkVersion, cVersionParts);
+    }
+
+    int iMajor = 0;
+    int iMinor = 0;
+    int iPatch = 0;
```

Comment:
These aren't used and can be deleted, right?

Reply:
I've removed these